## Hacker News Posts
---
The data set used for this project consists of Hacker News Posts.

The questions we'd like to solve are:

1. Between ```Ask HN``` and ```Show HN```, which type of post receives the most comments on average?
2. Does the time of which the post is created determine the average comments received?

In [5]:
from csv import reader

file = open("hacker_news.csv")
file_contents = reader(file)
hn = list(file_contents)

In [8]:
# print first 5 rows
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [11]:
# save header
headers = hn[0]

In [12]:
# remove header from hn[]
hn = hn[1:]

In [13]:
# make sure the header has been retrieved
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [14]:
# make sure the header has successfully been removed
print(hn[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


### Filter Data

In [22]:
ask_posts, show_posts, other_posts = [], [], []

for row in hn:
    title = row[1]
    title = title.lower()
    
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [18]:
# check number of posts
total_ask_posts = len(ask_posts)
total_show_posts = len(show_posts)
total_other_posts = len(other_posts)

In [23]:
print(total_ask_posts)
print(total_show_posts)
print(total_other_posts)

1744
1162
17194


In [24]:
# print first 5 rows of each list
print(ask_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [25]:
print(show_posts[:5])

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


#### Question I: 
Do ```Ask Posts``` or ```Show Posts``` receive more comments on average?

In [26]:
# Avg ask comments
total_ask_comments = 0
for row in ask_posts:
    num_comments = row[4]
    total_ask_comments += int(num_comments)
    
avg_ask_comments = total_ask_comments / total_ask_posts

In [35]:
print("Avg Ask Comments: ", avg_ask_comments)

Avg Ask Comments:  14.038417431192661


In [30]:
# Avg show comments
total_show_comments = 0
for row in show_posts:
    num_comments = row[4]
    total_show_comments += int(num_comments)
    
avg_show_comments = total_show_comments / total_show_posts

In [33]:
print("Avg Show Comments: ", avg_show_comments)

Avg Show Comments:  10.31669535283993


#### Observation:
On average, according to our sample data, ```ask``` posts tend to receive more comments compared to ```show``` posts. Since ```ask``` posts are more likely to receive more comments, we will focus the rest of our analysis on these posts.

#### Question II: At what time are ```ask``` posts more likely to attract comments?

In [48]:
import datetime as dt

In [60]:
result_list = []
for row in ask_posts:
    result_list.append(
        [row[6], 
         int(row[4])
        ])

In [62]:
counts_by_hour, comments_by_hour = {}, {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    # extract date
    date = row[0]
    # extract comment number
    comment = row[1]
    
    #format date and extract the hour from that date
    time = dt.datetime.strptime(date, date_format).strftime("%H")
     
    if time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment
    

In [64]:
counts_by_hour

{'00': 55,
 '01': 60,
 '02': 58,
 '03': 54,
 '04': 47,
 '05': 46,
 '06': 44,
 '07': 34,
 '08': 48,
 '09': 45,
 '10': 59,
 '11': 58,
 '12': 73,
 '13': 85,
 '14': 107,
 '15': 116,
 '16': 108,
 '17': 100,
 '18': 109,
 '19': 110,
 '20': 80,
 '21': 109,
 '22': 71,
 '23': 68}

### Average comment per hour

In [67]:
# average number of comments for posts created during each hour
avg_by_hour = []

for hour in counts_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour]])

In [68]:
avg_by_hour

[['17', 11.46],
 ['04', 7.170212765957447],
 ['02', 23.810344827586206],
 ['08', 10.25],
 ['12', 9.41095890410959],
 ['19', 10.8],
 ['20', 21.525],
 ['06', 9.022727272727273],
 ['03', 7.796296296296297],
 ['13', 14.741176470588234],
 ['23', 7.985294117647059],
 ['00', 8.127272727272727],
 ['18', 13.20183486238532],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['11', 11.051724137931034],
 ['05', 10.08695652173913],
 ['09', 5.5777777777777775],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['15', 38.5948275862069],
 ['10', 13.440677966101696],
 ['21', 16.009174311926607],
 ['07', 7.852941176470588]]

### Sort lists and print five(5) highest values

In [70]:
swap_avg_by_hour = []
# swap column
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
swap_avg_by_hour

[[11.46, '17'],
 [7.170212765957447, '04'],
 [23.810344827586206, '02'],
 [10.25, '08'],
 [9.41095890410959, '12'],
 [10.8, '19'],
 [21.525, '20'],
 [9.022727272727273, '06'],
 [7.796296296296297, '03'],
 [14.741176470588234, '13'],
 [7.985294117647059, '23'],
 [8.127272727272727, '00'],
 [13.20183486238532, '18'],
 [13.233644859813085, '14'],
 [16.796296296296298, '16'],
 [11.051724137931034, '11'],
 [10.08695652173913, '05'],
 [5.5777777777777775, '09'],
 [11.383333333333333, '01'],
 [6.746478873239437, '22'],
 [38.5948275862069, '15'],
 [13.440677966101696, '10'],
 [16.009174311926607, '21'],
 [7.852941176470588, '07']]

In [72]:
# sort the above list
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [77]:
print("Top 5 Hours for Ask Post Comments")

for avg, hour in sorted_swap[:5]:
    convert_hour_to_mst = int(hour)-2
    convert_hour_to_mst = str(convert_hour)
    
    print(
    "{} EST ({} MST): {:.2f} average comments per post".format(dt.datetime.strptime(hour, "%H").strftime("%H:%M"),
                                                              dt.datetime.strptime(convert_hour_to_mst, "%H").strftime("%H:%M"),
                                                              avg)
    )

Top 5 Hours for Ask Post Comments
15:00 EST (13:00 MST): 38.59 average comments per post
02:00 EST (00:00 MST): 23.81 average comments per post
20:00 EST (18:00 MST): 21.52 average comments per post
16:00 EST (14:00 MST): 16.80 average comments per post
21:00 EST (19:00 MST): 16.01 average comments per post


#### Observation
To increase your chances of receiving comments, the hour with the highest chance is 13:00 MST with an average of 38.59 comments per post. This is a 62% increase compared to the second highest time to post (Midnight).

### Conclusion

Of the posts that contained comments and were randomly chosen for this data set, we determined the type of posts receiving the most comments on average were ```Ask HN``` posts compared to ```Show HN``` posts. We also discovered that there is a correlation between the time a user posts their question and the number of comments they may receive. 

On average, an ask post will receive 14 comments. Yet, our findings illustrate that at 13:00 MST this average increases to approximately 38 comments per post. The second highest time is at midnight, however, the number of comments received is 60% lower than the highest time to post.

With this analysis, we reccomend posting your questions at 13:00 MST to maximize the number of responses you'd like to receive.